cpp/arcticdb/version/version_core.cpp

```diff
@@ -890,25 +890,18 @@ bool read_incompletes_to_pipeline(
         ensure_timeseries_norm_meta(*pipeline_context->norm_meta_, pipeline_context->stream_id_, sparsify);
     }
 
-    if (!dynamic_schema) {
-        const auto first_different_seg = std::adjacent_find(
```

Comment:
This check was the cause of the high memory use. It calls `.segment()` on every `SliceAndKey` which loads the segment from storage, and then holds it live in memory for the rest of the processing pipeline.

This PR moves this check so that it is done inline during the aggregation of the incomplete segments during compaction.

I also implement changes to make sure that we clean up any data segments that we wrote during a compaction that eventually failed. This change also applies to `defragment_symbol_data` since it also uses the `do_compact` function. This seems correct.